### PR TITLE
Add ADAMAX updater impl.

### PR DESCRIPTION
**Pull request by crockpotveggies**
## What changes were proposed in this pull request?

Adds the ADAMAX updater implementation.

screen shot 2017-05-05 at 6 17 57 pm

## How was this patch tested?

Jackson dependency issues have prevented me from testing this. Asking for help on that front.

Please review
https://github.com/deeplearning4j/deeplearning4j/blob/master/CONTRIBUTING.md before opening a pull request.

### CODE REVIEW

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/AdaMax.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+     */
+    @Override
+    public INDArray getGradient(INDArray gradient, int iteration) {
+        if (m == null || u == null)
+            throw new IllegalStateException("Updater has not been initialized with view state");
+
+        INDArray oneMinusBeta1Grad = gradient.mul(1.0 - beta1);
+        m.muli(beta1).addi(oneMinusBeta1Grad);
+
+        u.assign(Transforms.max(u.muli(beta2), Transforms.abs(gradient)));
```
Comment (AlexDBlack) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/AdaMax.java]: I think we can do the following, and avoid one of the working arrays + an assign op
```
u.muli(beta2);
Transforms.abs(gradient,false);   //In-place should be OK here, original gradient values aren't used again later
Nd4j.getExecutioner().exec(new Max(u,gradient,u))
```

File: nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/AdaMax.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+        double beta1t = FastMath.pow(beta1, iteration + 1);
+        double beta2t = FastMath.pow(beta2, iteration + 1);
+
+        double alphat = learningRate * FastMath.sqrt(1 - beta2t) / (1 - beta1t);
```
Comment (AlexDBlack) [inline nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/learning/AdaMax.java]: Shouldn't this just be ```learningRate / (1-beta1t)```?

**Comment by crockpotveggies** [comment]
@AlexDBlack updated for your reviewing pleasure.

**Comment by crockpotveggies** [comment]
After a quick chat with @AlexDBlack this looks to be in good shape. Merging this so that we can ensure changes are picked up by https://github.com/deeplearning4j/nd4j/pull/1819

**Pull request merged**